Allow git clone instead of tarball

I'm working on [micro-editor.github.io](https://github.com/magnusviri/tea_pantry.extra/blob/micro-editor.github.io/projects/micro-editor.github.io/package.yml) and this is the output when I build.

```
> ./pantry.core/scripts/build.ts micro-editor.github.io
{ ignoring: "1.0rc1", reason: "unparsable" }
{ ignoring: "v1.0rc2", reason: "unparsable" }
{ ignoring: "v2.0.0-rc1", reason: "unparsable" }
{ ignoring: "v2.0.0-rc2", reason: "unparsable" }
{ ignoring: "v2.0.0-rc3", reason: "unparsable" }
{ ignoring: "nightly", reason: "unparsable" }
fetching micro-editor.github.io@2.0.11
{ ignoring: "1.0rc1", reason: "unparsable" }
{ ignoring: "v1.0rc2", reason: "unparsable" }
{ ignoring: "v2.0.0-rc1", reason: "unparsable" }
{ ignoring: "v2.0.0-rc2", reason: "unparsable" }
{ ignoring: "v2.0.0-rc3", reason: "unparsable" }
{ ignoring: "nightly", reason: "unparsable" }
querying
cache: hit
+ cd /opt/tea/micro-editor.github.io/src/v2.0.11
+ export SRCROOT=/opt/tea/micro-editor.github.io/src/v2.0.11
+ SRCROOT=/opt/tea/micro-editor.github.io/src/v2.0.11
+ export PATH=/opt/tea/tea.xyz/gx/cc/v0.1.1/bin:/opt/tea/openssl.org/v1.1.118/bin:/opt/tea/go.dev/v1.19.2/bin:/Users/james/.tea_bin:/usr/bin:/bin:/usr/sbin:/sbin
+ PATH=/opt/tea/tea.xyz/gx/cc/v0.1.1/bin:/opt/tea/openssl.org/v1.1.118/bin:/opt/tea/go.dev/v1.19.2/bin:/Users/james/.tea_bin:/usr/bin:/bin:/usr/sbin:/sbin
+ export PKG_CONFIG_PATH=/opt/tea/openssl.org/v1.1.118/lib/pkgconfig
+ PKG_CONFIG_PATH=/opt/tea/openssl.org/v1.1.118/lib/pkgconfig
+ export LIBRARY_PATH=/opt/tea/openssl.org/v1.1.118/lib:/opt/tea/go.dev/v1.19.2/lib
+ LIBRARY_PATH=/opt/tea/openssl.org/v1.1.118/lib:/opt/tea/go.dev/v1.19.2/lib
+ export LD_LIBRARY_PATH=/opt/tea/openssl.org/v1.1.118/lib:/opt/tea/go.dev/v1.19.2/lib
+ LD_LIBRARY_PATH=/opt/tea/openssl.org/v1.1.118/lib:/opt/tea/go.dev/v1.19.2/lib
+ export CPATH=/opt/tea/openssl.org/v1.1.118/include
+ CPATH=/opt/tea/openssl.org/v1.1.118/include
+ export DYLD_FALLBACK_LIBRARY_PATH=/opt/tea/openssl.org/v1.1.118/lib:/opt/tea/go.dev/v1.19.2/lib
+ DYLD_FALLBACK_LIBRARY_PATH=/opt/tea/openssl.org/v1.1.118/lib:/opt/tea/go.dev/v1.19.2/lib
+ export SSL_CERT_FILE=/opt/tea/openssl.org/v1.1.118/ssl/cert.pem
+ SSL_CERT_FILE=/opt/tea/openssl.org/v1.1.118/ssl/cert.pem
+ export LDFLAGS=-Wl,-rpath,/opt/tea
+ LDFLAGS=-Wl,-rpath,/opt/tea
+ export TEA_PREFIX=/opt/tea
+ TEA_PREFIX=/opt/tea
+ export MACOSX_DEPLOYMENT_TARGET=11.0
+ MACOSX_DEPLOYMENT_TARGET=11.0
+ export PATH=/Users/james/Tea/pantry.core/scripts/brewkit:/opt/tea/tea.xyz/gx/cc/v0.1.1/bin:/opt/tea/openssl.org/v1.1.118/bin:/opt/tea/go.dev/v1.19.2/bin:/Users/james/.tea_bin:/usr/bin:/bin:/usr/sbin:/sbin
+ PATH=/Users/james/Tea/pantry.core/scripts/brewkit:/opt/tea/tea.xyz/gx/cc/v0.1.1/bin:/opt/tea/openssl.org/v1.1.118/bin:/opt/tea/go.dev/v1.19.2/bin:/Users/james/.tea_bin:/usr/bin:/bin:/usr/sbin:/sbin
+ make build-tags
git fetch --tags
fatal: not a git repository (or any of the parent directories): .git
make: *** [fetch-tags] Error 128
error: Uncaught Error: cmd failed: 2: /opt/tea/micro-editor.github.io/src/build.sh
    if (!exit.success) throw new RunError(exit.code, cmd)
                             ^
    at run (file:///Users/james/Tea/pantry.core/src/utils/index.ts:234:30)
    at async build (file:///Users/james/Tea/pantry.core/scripts/build/build.ts:113:5)
    at async __build (file:///Users/james/Tea/pantry.core/scripts/build/build.ts:34:24)
    at async _build (file:///Users/james/Tea/pantry.core/scripts/build/build.ts:21:12)
    at async file:///Users/james/Tea/pantry.core/scripts/build.ts:43:3
```

The build script for micro requires the download to be a git repo. Is there a way to do a git clone instead of downloading a tarball?

--- a/lib/usePantry.ts
+++ b/lib/usePantry.ts
@@ -102,10 +102,24 @@ const getRawDistributableURL = (yml: PlainObject) => {
   }
 }
 
+const getGitDistribution = ({ pkg, git, ref }: { pkg: Package, git: string, ref: string }) => {
+  const moustaches = useMoustaches()
+
+  const ref_ = moustaches.apply(ref, [
+    ...moustaches.tokenize.version(pkg.version),
+    ...moustaches.tokenize.host()
+  ])
+
+  return { url: new URL(git), ref: ref_, stripComponents: 0, type: 'git' }
+}
+
 const getDistributable = async (pkg: Package) => {
   const moustaches = useMoustaches()
 
   const yml = await entry(pkg).yml()
+
+  if (yml.distributable?.git) { return getGitDistribution({ pkg, ...yml.distributable}) }
+
   let urlstr = getRawDistributableURL(yml)
   if (!urlstr) return
   let stripComponents: number | undefined
@@ -120,7 +134,7 @@ const getDistributable = async (pkg: Package) => {
 
   const url = new URL(urlstr)
 
-  return { url, stripComponents }
+  return { url, stripComponents, type: 'url' }
 }
 
 const getScript = async (pkg: Package, key: 'build' | 'test', deps: Installation[]) => {

--- a/lib/usePantry.ts
+++ b/lib/usePantry.ts
@@ -134,7 +134,7 @@ const getDistributable = async (pkg: Package) => {
 
   const url = new URL(urlstr)
 
-  return { url, stripComponents, type: 'url' }
+  return { url, ref: undefined, stripComponents, type: 'url' }
 }
 
 const getScript = async (pkg: Package, key: 'build' | 'test', deps: Installation[]) => {

--- a/lib/usePantry.ts
+++ b/lib/usePantry.ts
@@ -103,6 +103,10 @@ const getRawDistributableURL = (yml: PlainObject) => {
 }
 
 const getGitDistribution = ({ pkg, git, ref }: { pkg: Package, git: string, ref: string }) => {
+  if (!git.startsWith("git+")) throw new Error(`invalid git url; explicitly use git+https:// or git+ssh://: ${git}`)
+
+  const url = new URL(git.replace(/^git\+http/, 'http'))
+
   const moustaches = useMoustaches()
 
   const ref_ = moustaches.apply(ref, [
@@ -110,7 +114,7 @@ const getGitDistribution = ({ pkg, git, ref }: { pkg: Package, git: string, ref:
     ...moustaches.tokenize.host()
   ])
 
-  return { url: new URL(git), ref: ref_, stripComponents: 0, type: 'git' }
+  return { url, ref: ref_, stripComponents: 0, type: 'git' }
 }
 
 const getDistributable = async (pkg: Package) => {

--- a/libexec/fetch.ts
+++ b/libexec/fetch.ts
@@ -5,13 +5,16 @@ args:
   - deno
   - run
   - --allow-net
-  - --allow-run=curl
+  - --allow-run=curl,git,tar
   - --allow-read
   - --allow-write
   - --allow-env
   - --unstable
 dependencies:
   curl.se: '*'
+  git-scm.org: '*'
+  gnu.org/tar: '*'
+  tukaani.org/xz: '*'
 ---*/
 
 //TODO verify the sha
@@ -40,7 +43,7 @@ const { flags, unknown: [pkgname] } = parseFlags(Deno.args, {
 })
 
 const pkg = await pantry.resolve(parse(pkgname))
-const { url } = await pantry.getDistributable(pkg) ?? {}
+const { url, ref, type } = await pantry.getDistributable(pkg) ?? {}
 
 if (!url) {
   console.error("warn: pkg has no srcs: ", pkg)
@@ -49,7 +52,12 @@ if (!url) {
 
 try {
   const zipfile = await (async () => {
-    const stowage: Stowage = { pkg, type: 'src', extname: url.path().extname() }
+    const stowage: Stowage = (() => {
+      if (type === "git") {
+        return { pkg, type: 'src', extname: `.tar.xz` }
+      }
+      return { pkg, type: 'src', extname: url.path().extname() }
+    })()
 
     const dst = (() => {
       if (flags.outputDir) {
@@ -59,6 +67,8 @@ try {
         return Path.cwd().join(flags.o ?? panic())
       }
     })()
+
+    if (type === "git") return clone({ dst, src: url, ref})
 
     try {
       // first try the original location
@@ -98,6 +108,48 @@ async function download({ dst, src }: { dst: Path, src: URL }) {
   } else {
     // locally using our download function as it knows how to cache properly
     await useDownload().download({ dst, src })
+  }
+  return dst
+}
+
+async function clone({ dst, src, ref }: { dst: Path, src: URL, ref: string }) {
+  const tmp = Path.mktemp({})
+  const proc = new Deno.Command("git", {
+    args: [
+      "clone",
+      "--quiet",
+      "--depth=1",
+      "--branch",
+      ref,
+      src.toString(),
+      tmp.string,
+    ],
+    // `git` uses stderr for... non errors, and --quiet
+    // doesn't touch them
+    stderr: "null",
+  })
+  const status = await proc.spawn().status
+  if (!status.success) {
+    console.error({ dst, src })
+    throw new Error(`git failed to clone ${src}`)
+  }
+
+  const proc2 = new Deno.Command("tar", {
+    args: [
+      "-C",
+      tmp.string,
+      // Prevents `tar: Ignoring unknown extended header keyword 'LIBARCHIVE.xattr.com.apple.provenance'`
+      // when unpacking on darwin
+      "--no-xattrs",
+      "-czf",
+      dst.string,
+      ".",
+    ]
+  })
+  const status2 = await proc2.spawn().status
+  if (!status2.success) {
+    console.error({ dst, src })
+    throw new Error(`tar failed to create ${dst}`)
   }
   return dst
 }

--- a/libexec/fetch.ts
+++ b/libexec/fetch.ts
@@ -112,8 +112,13 @@ async function download({ dst, src }: { dst: Path, src: URL }) {
   return dst
 }
 
+// Clones a git repo, then builds a src tarball from it
+// This allows our system to treat git repos as if they were
+// tarballs, improving internal consistency
 async function clone({ dst, src, ref }: { dst: Path, src: URL, ref: string }) {
   const tmp = Path.mktemp({})
+
+  // Clone the specific ref to our temp dir
   const proc = new Deno.Command("git", {
     args: [
       "clone",
@@ -134,6 +139,7 @@ async function clone({ dst, src, ref }: { dst: Path, src: URL, ref: string }) {
     throw new Error(`git failed to clone ${src}`)
   }
 
+  // Create a tarball from the temp dir
   const proc2 = new Deno.Command("tar", {
     args: [
       "-C",

--- a/libexec/fetch.ts
+++ b/libexec/fetch.ts
@@ -68,7 +68,7 @@ try {
       }
     })()
 
-    if (type === "git") return clone({ dst, src: url, ref})
+    if (type === "git") return clone({ dst, src: url, ref: ref! })
 
     try {
       // first try the original location

--- a/libexec/query.ts
+++ b/libexec/query.ts
@@ -45,12 +45,16 @@ const version = pkg.constraint.single() ?? panic()
 pkg = {project: pkg.project, version }
 
 if (src) {
-  const { url } = await usePantry().getDistributable(pkg) ?? {}
+  const { url, type } = await usePantry().getDistributable(pkg) ?? {}
   if (!url) {
     console.error("warn: pkg has no srcs: ", str(pkg))
     Deno.exit(0)  // NOT AN ERROR EXIT CODE THO
   }
-  const stowage: Stowage = { pkg, type: 'src', extname: url.path().extname() }
+  const stowage: Stowage = (() => {
+    if (type === 'git')
+      return { pkg, type: 'src', extname: '.tar.xz' }
+    return { pkg, type: 'src', extname: url.path().extname() }
+  })()
   const path = flatmap(Deno.env.get("SRCROOT"), x => new Path(x))
   const cache_path = useCache().path(stowage)
   if (path?.join("projects").isDirectory()) {

--- a/share/TEMPLATE.pkg.yml
+++ b/share/TEMPLATE.pkg.yml
@@ -16,6 +16,10 @@ distributable:
   # `1`. If after extraction the src-directory is funky try `0` or `2`.
   strip-components: 1
   # alternatively, we can use direct git checkouts from refs
+  # but we should prefer official source distributions.
+  # often CI/CD will do some post-processing of the source
+  # for release, including setting version numbers, running
+  # autotools, etc.
   url: https://github.com/user/repo
   ref: v{{version}}}
 

--- a/share/TEMPLATE.pkg.yml
+++ b/share/TEMPLATE.pkg.yml
@@ -20,7 +20,7 @@ distributable:
   # often CI/CD will do some post-processing of the source
   # for release, including setting version numbers, running
   # autotools, etc.
-  url: https://github.com/user/repo
+  url: git+https://github.com/user/repo
   ref: v{{version}}}
 
 # if there’s a github then we can parse the versions

--- a/share/TEMPLATE.pkg.yml
+++ b/share/TEMPLATE.pkg.yml
@@ -15,6 +15,9 @@ distributable:
   # extracted tarball. It only works with tarballs. You almost always want
   # `1`. If after extraction the src-directory is funky try `0` or `2`.
   strip-components: 1
+  # alternatively, we can use direct git checkouts from refs
+  url: https://github.com/user/repo
+  ref: v{{version}}}
 
 # if there’s a github then we can parse the versions
 versions:

